pilot/pkg/networking/core/v1alpha3/listener_waypoint.go

```diff
@@ -301,127 +225,120 @@ func (lb *ListenerBuilder) buildWaypointInboundVIP(svcs map[host.Name]*model.Ser
-			if port.Protocol == protocol.UDP {
-				continue
-			}
-			cc := inboundChainConfig{
-				clusterName: model.BuildSubsetKey(model.TrafficDirectionInboundPod, "", host.Name(wl.PodIP), port.Port),
-				port: ServiceInstancePort{
-					Name:       port.Name,
-					Port:       uint32(port.Port),
-					TargetPort: uint32(port.Port),
-					Protocol:   port.Protocol,
+	{
+		// Direct pod access chain.
+		directChain := &listener.FilterChain{
+			Name: "direct",
+			Filters: []*listener.Filter{{
+				Name: wellknown.TCPProxy,
+				ConfigType: &listener.Filter_TypedConfig{
+					TypedConfig: protoconv.MessageToAny(&tcp.TcpProxy{
+						StatPrefix:       "direct",
+						ClusterSpecifier: &tcp.TcpProxy_Cluster{Cluster: "encap"},
+					}),
 				},
-				bind:  "0.0.0.0",
-				hbone: true,
-			}
-			name := cc.clusterName
-
-			tcpName := name + "-tcp"
-			tcpChain := &listener.FilterChain{
-				Filters: wlBuilder.buildInboundNetworkFilters(cc),
-				Name:    tcpName,
-			}
-
-			httpName := name + "-http"
-			httpChain := &listener.FilterChain{
-				Filters: wlBuilder.buildInboundNetworkFiltersForHTTP(cc),
-				Name:    httpName,
-			}
-			l := &listener.Listener{
-				Name:              name,
-				ListenerSpecifier: &listener.Listener_InternalListener{InternalListener: &listener.Listener_InternalListenerConfig{}},
-				ListenerFilters:   []*listener.ListenerFilter{util.InternalListenerSetAddressFilter()},
-				TrafficDirection:  core.TrafficDirection_INBOUND,
-				FilterChains:      []*listener.FilterChain{},
-			}
-			if port.Protocol.IsUnsupported() {
-				// If we need to sniff, insert two chains and the protocol detector
-				l.FilterChains = append(l.FilterChains, tcpChain, httpChain)
-				l.FilterChainMatcher = match.NewAppProtocol(match.ProtocolMatch{
-					TCP:  match.ToChain(tcpName),
-					HTTP: match.ToChain(httpName),
-				})
-			} else if port.Protocol.IsHTTP() {
-				// Otherwise, just insert HTTP/TCP
-				l.FilterChains = append(l.FilterChains, httpChain)
-			} else {
-				l.FilterChains = append(l.FilterChains, tcpChain)
-			}
-			listeners = append(listeners, l)
+			}},
+		}
+		// Workload IP filtering happens here.
+		ipRange := []*xds.CidrRange{}
+		for _, wlx := range wls {
+			cidr := util.ConvertAddressToCidr(wlx.WorkloadInfo.PodIP)
+			ipRange = append(ipRange, &xds.CidrRange{
+				AddressPrefix: cidr.AddressPrefix,
+				PrefixLen:     cidr.PrefixLen,
+			})
 		}
+		chains = append(chains, directChain)
+		ipMatcher.RangeMatchers = append(ipMatcher.RangeMatchers,
+			&matcher.IPMatcher_IPRangeMatcher{
+				Ranges:  ipRange,
+				OnMatch: match.ToChain(directChain.Name),
+			})
 	}
-	return listeners
+	l := &listener.Listener{
+		Name:              "internal",
+		ListenerSpecifier: &listener.Listener_InternalListener{InternalListener: &listener.Listener_InternalListenerConfig{}},
+		ListenerFilters: []*listener.ListenerFilter{
+			util.InternalListenerSetAddressFilter(),
+			xdsfilters.HTTPInspector,
```

Comment:
Do we need TLS inspector as well?

I think long term this will be problematic, as we will sniff all traffic, even explicitly declare, which breaks server-first. But I have no concern merging this for now

Reply:
Ideally we could enable it selectively along the matcherTree but not sure how crazy that would be to implement in Envoy

Reply:
There's a way to skip the inspectors. It sounds like most of the time we handle HTTP, so we should first find out when we should skip HTTP inspector and then mark those routes/IP appropriately. It could be a per endpoint-port annotations.

I don't think we need TLS inspector. We always have TLS on HBONE.

Reply:
TLS inspector: I would think it would be on the inner body not the HBONE? Although its a really niche match, I think it would be "match service X, and then ensure SNI is X" in our API, as our API is pretty restrictive in TLSRoutes. I have no problem skipping this for now, but we should probably open an issue to track